okhttp/src/main/java/okhttp3/internal/ws/WebSocketWriter.java

```diff
@@ -221,7 +241,13 @@ void writeMessageFrame(int formatOpcode, long byteCount, boolean isFirstFrame,
     @Override public void write(Buffer source, long byteCount) throws IOException {
       if (closed) throw new IOException("closed");
 
-      buffer.write(source, byteCount);
+      Buffer uncompressedSource = source;
+      if (messageDeflater != null) {
+        uncompressedSource = CompressionCorrector.applyPostDeflate(
+            messageDeflater.deflate(source));
+      }
+
+      buffer.write(uncompressedSource, uncompressedSource.size());
```

Comment:
Just realized that compressing the whole message here, inside FrameSink.write() is not correct, as the method guarantees to write exactly `byteCount` number of bytes. 

Besides, there is logic in `RealWebSocket.send(ByteString data, int formatOpcode)` which checks for queue overflow. With this change, the logic there will count the number of _uncompressed_ bytes, which seems to be incorrect as it differs from the number of bytes to be sent on the wire. 
So, ideally the message data should be compressed by the time it is put in `messageAndCloseQueue` in the above mentioned `send` method. However, the method does not run on the writer thread, and we don't know the thread it is running on.  

Thinking how to resolve this... Suggestions are welcome!

Reply:
If we put aside the queue overflow checks, we could put the deflation code in the `RealWebSocket.writeOneFrame` method, like that: 

```
ByteString data = ((Message) messageOrClose).data;

// deflate data here and pass it down to newMessageSink

BufferedSink sink = Okio.buffer(writer.newMessageSink(
    ((Message) messageOrClose).formatOpcode, data.size()));
sink.write(data);
sink.close();
```

However, it does not seem to be a good solution.

Reply:
On second thought, `MAX_QUEUE_SIZE` is 16MB. I suppose it could be 12MB or 20MB or any other reasonable number. So if we treat `queueSize` as the number of _uncompressed_ bytes in the queue, the above solution with `RealWebSocket.writeOneFrame` could work.

Reply:
Implemented this approach.

Reply:
Good call.